Content.Server/Explosion/EntitySystems/SupermatterGrenadeSystem.cs

```diff
@@ -0,0 +1,97 @@
+    {
+        if(TryComp<SingularityDistortionComponent>(uid, out var singulo))
+            component.Distortion = singulo;
+        if (TryComp<GravityWellComponent>(uid, out var well))
+            component.GravityWell = well;
+    }
+    private void OnExplode(EntityUid uid, SupermatterGrenadeComponent component, TriggerEvent args)
+    {
+        if (component.IsExploded)
+        {
+            if (component.IsGravityPulling && component.GravityPullEndSound != null)
+                _audio.PlayPvs(component.GravityPullEndSound, uid,
+                    AudioParams.Default.WithVolume(component.GravityPullEndSoundVolume));
+            return;
+        }
+        _container.TryRemoveFromContainer(uid, true);
+        if (component.AnchorOnGravityPull)
+            _transformSystem.AnchorEntity(uid, Transform(uid));
+        var param = AudioParams.Default;
+        if (component.GravityPullStartSound != null)
+            _audio.PlayPvs(component.GravityPullStartSound, uid,
+                AudioParams.Default.WithVolume(component.GravityPullStartSoundVolume));
+        if (HasComp<PointLightComponent>(uid))
+            _pointLightSystem.SetEnabled(uid, true);
+        if (component.Distortion != null)
+            component.Distortion.Intensity = component.DistortionIntensity;
+        if (component.GravityWell != null)
+            component.GravityWell.BaseRadialAcceleration = component.BaseRadialAcceleration;
+
+        component.IsGravityPulling = true;
+        component.ExplosionWillOccurIn =
+            _timing.CurTime + TimeSpan.FromSeconds(component.TimeTillExplosion);
+        component.GravityPullWillOccurIn =
+            _timing.CurTime + TimeSpan.FromSeconds(component.GravityPullLoopSoundOffset);
+        if (!HasComp<ExplodeOnTriggerComponent>(uid) && !HasComp<DeleteOnTriggerComponent>(uid))
+        {
+            if (component.ExplodeAfterGravityPull)
+                _entityManager.AddComponent<ExplodeOnTriggerComponent>(uid);
+            else
+                _entityManager.AddComponent<DeleteOnTriggerComponent>(uid);
+        }
+    }
+
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+
+        var enumerator = EntityQueryEnumerator<SupermatterGrenadeComponent>();
+        while (enumerator.MoveNext(out var uid, out var component))
+        {
+            if (!component.IsGravityPulling)
+                continue;
+            if (!component.IsGravitySoundBegan && component.GravityPullWillOccurIn <= _timing.CurTime)
+            {
+                component.IsGravitySoundBegan = true;
+                if (TryComp<AmbientSoundComponent>(uid, out var ambience))
+                    _ambient.SetAmbience(uid, true, ambience);
+            }
+            if (!component.IsExploded && component.ExplosionWillOccurIn <= _timing.CurTime)
+            {
+                component.IsExploded = true;
+                _triggerSystem.Trigger(uid);
+            }
+        }
```

Comment:
Isn't this just duplicating existing grenade functionality? Components and systems should be generic and re-useable otherwise it makes a codebase of this size a maintenance nightmare. I'm pretty sure we already have timer triggers.